jdk/test/java/security/SignedJar/spi-calendar-provider/TestSPISigned.java

```diff
@@ -0,0 +1,136 @@
+ * @bug 8297684 8269039
+ * @summary Checking custom CalendarDataProvider with SPI contained in signed jar does
+ *          not produce NPE.
+ * @library /lib/testlibrary
+ * @library provider
+ * @build baz.CalendarDataProviderImpl
+ * @run main/timeout=600 TestSPISigned
+ */
+public class TestSPISigned {
+
+    private static final String TEST_CLASSES = System.getProperty("test.classes", ".");
+    private static final String TEST_SRC = System.getProperty("test.src", ".");
+
+    private static final Path META_INF_DIR = Paths.get(TEST_SRC, "provider", "meta");
+    private static final Path META_INF_DIR_SOURCE = META_INF_DIR.resolve("META-INF");
+    private static final Path SERVICES_DIR_SOURCE = META_INF_DIR_SOURCE.resolve("services");
+    private static final Path PROVIDER_SOURCE = SERVICES_DIR_SOURCE.resolve("java.util.spi.CalendarDataProvider");
+    private static final Path PROVIDER_PARENT = Paths.get(TEST_CLASSES);
+    private static final Path PROVIDER_DIR = PROVIDER_PARENT.resolve("provider");
+    private static final Path META_INF_DIR_TARGET = PROVIDER_DIR.resolve("META-INF");
+    private static final Path SERVICES_DIR_TARGET = META_INF_DIR_TARGET.resolve("services");
+    private static final Path PROVIDER_TARGET = SERVICES_DIR_TARGET.resolve("java.util.spi.CalendarDataProvider");
+    private static final Path MODS_DIR = Paths.get("mods");
+    private static final Path UNSIGNED_JAR = MODS_DIR.resolve("unsigned-with-locale.jar");
+    private static final Path SIGNED_JAR = MODS_DIR.resolve("signed-with-locale.jar");
+
+    public static void main(String[] args) throws Throwable {
+        if (args.length == 1) {
+            String arg = args[0];
+            if ("run-test".equals(arg)) {
+                System.out.println("Debug: Running test");
+                String provProp = System.getProperty("java.ext.dirs");
+                String[] extDirs = provProp.split(File.pathSeparator);
+                boolean extDirFound = false;
+                for (int i = 0; i < extDirs.length; i++) {
+                    if (extDirs[i].equals(MODS_DIR.toAbsolutePath().toString())) {
+                        extDirFound = true;
+                        break;
+                    }
+                }
+                if (!extDirFound) {
+                   throw new RuntimeException("Test failed! Expected -Djava.ext.dirs to include CalendarDataProvider jar");
+                }
+                doRunTest();
+            } else {
+               throw new RuntimeException("Test failed! Expected 'run-test' arg for test run");
+            }
+        } else {
+            // Set up signed jar with custom calendar data provider
+            //
+            // 1. Create jar with custom CalendarDataProvider
+            Files.createDirectory(META_INF_DIR_TARGET);
+            Files.createDirectory(SERVICES_DIR_TARGET);
+            Files.copy(PROVIDER_SOURCE, PROVIDER_TARGET);
+            JarUtils.createJarFile(UNSIGNED_JAR, PROVIDER_DIR);
```

Comment:
I take it the JAR file does not need to be updated with the META-INF directory (11u has `JarUtils.updateJarFile(UNSIGNED_JAR, META_INF_DIR);`) because it is instead located via `-Djava.ext.dirs`?

Reply:
This has to do with different test libraries in place in 8u. `JarUtils.createJarFile()` here creates a jar file based on a directory (in this case the `test.classes` dir), where we already have the `*.class` files and we copy the `META-INF` folder from the test sources there. The javadoc of `createJarFile()` has:

```
Creates a JAR file from the contents of a directory.

Equivalent to {@code jar cf <jarfile> -C <dir> .}
```

So we create the right jar file with the META-INF directory included right from the start (as compared to the 11u version, where a jar file is being created *without* `META-INF/services` files and then updated with `META-INF` bits).